### DISCUSSION TITLE
Colors wrong when converting YUV-type SVS files with dzsave

### DISCUSSION THREAD

**Discussion by JRhodes-microscope** [libvips/libvips#3116] [Q&A] [answered]
This note/question is regarding a problem with color representation when dzsave() is used to convert "YUV" SVS files to Google-Maps-like virtual slides.  Before trying to find a solution ("reinventing-the-wheel"), I wanted to ask:  **_has anyone already has encountered & overcome this problem?_**

I have been using the dzsave function to convert SVS-formatted virtual slides into the Google-Maps file architecture, i.e., using the command: "**vips dzsave** _SVS-filename_ _GoogleMap-root-directory_ **--layout google --background 0 --centre**".  Normally, this works beautifully (THANKS!!).  However, we recently moved the project to a server at Southern Illinois University Edwardsville (SIUE), and on this server, the colors are misrepresented in the jpeg files after conversion to Google-Maps format **if** the SVS file has the '_Image Type_'= "**SVS/J2K-YUV5**".  For instance, see: [UM-65-2 converted on the SIUE server](https://virtual-microscope.isg.siue.edu/NoSQLViewer?sb=4389&x=65000&y=65000&z=4&hidelogo) (https://virtual-microscope.isg.siue.edu/NoSQLViewer?sb=4389&x=65000&y=65000&z=4&hidelogo), which should look like: [UM-65-2 converted previously](https://virtual-microscope.isg.siue.edu/NoSQLViewer?sb=389&x=65000&y=65000&z=4&hidelogo) (https://virtual-microscope.isg.siue.edu/NoSQLViewer?sb=389&x=65000&y=65000&z=4&hidelogo).  The problem does NOT occur if the SVS file is of a different '_Image Type_' (such as "SVS JPEG2" or "SVS/JPEG2000 3").

According to my collaborator who is in charge of the SIUE server, the version of VIPS that he installed on the SIUE server in May of this year (2022) is: "libvips v8.10.5 11 latest/stable  samwiseg0", and it is running on "Oracle Linux Server release 8.6".  "ImageMagick-6.9.12.61-1.el8" also is installed, but I'm guessing that this may not be relevant since ImageMagick is not referenced when I invoke vips.  The color-representation problem only arose after we moved to SIUE; it did NOT occur when "YUV" SVS files were converted to Google-Maps format on the old server (at Pacific Northwest University, PNWU) which had VIPS version 8.4.5 installed, and it does NOT occur when my PNWU colleague converts the files on his Windows workstation using VIPS 8.12.2.

As background, our virtual microscope is [available on GitHub](https://github.com/MFPvirtual-microscope) (https://github.com/MFPvirtual-microscope), but please contact me before downloading the software since there hasn't been much interest in this software and the version on GitHub needs to be updated.  The impetus for developing this virtual microscope was to allow high-resolution, multifocal-plane virtual slides to be used in teaching Histology; for instance, on [UCSF67-273](https://virtual-microscope.isg.siue.edu/?sb=3012&x=148700&y=140100&z=9&hidelogo) (https://virtual-microscope.isg.siue.edu/?sb=3012&x=148700&y=140100&z=9&hidelogo), [UCSF172-485](https://virtual-microscope.isg.siue.edu/?sb=3003&x=97900&y=93300&z=9&f=4&hidelogo), or [UCSF32-70](https://virtual-microscope.isg.siue.edu/?sb=3010&x=97800&y=63500&z=9&f=19&hidelogo), click on the "cycle through focal planes" button on the Navigator.  It is intended that students would access the virtual slides through URL-links embedded in their lab handouts without having to navigate (or even know about) the database of virtual slides, but the faculty writing lab handouts probably would want to use the [SlideBox](https://virtual-microscope.isg.siue.edu/SlideBox/) (https://virtual-microscope.isg.siue.edu/SlideBox/) to search the database of available virtual slides.  The vast majority (>95%) of the virtual slides in the database currently are lower-resolution single-focal-plane virtual slides converted from legacy SVS-formatted slides, which were converted to Google-Maps format using dzsave().

Sorry that this note is so long.  Thanks for any help or suggestions you can provide!
    - - Jim Rhodes ([jrhodes98903@gmail.com]

**Comment by jcupitt** [comment, marked as answer]
Hi @JRhodes-microscope,

libvips uses openslide to load SVS files, so I would guess there is some problem with the openslide install on your SIUE server. What do you see when you run `vipsheader` on these slides? I see eg.:

```
$ vipsheader --version
vips-8.14.0
$ vipsheader CMU-1.svs
CMU-1.svs: 46000x32914 uchar, 4 bands, srgb, openslideload
```

So I'm running 8.14, and svs files are being loaded with `openslideload`. 

Do you see `tiffload` there? If you do, your libvips hasn't been configured with openslide.

**Comment by JRhodes-microscope** [comment]
John:

Many, many thanks!  I really appreciate your help ... and we all appreciate
libVips, which is tremendously useful software.

I ran the diagnostics you indicated:
$ vipsheader --version
vips-8.12.2-Tue Jan 25 09:34:32 UTC 2022

and (UVa-LY03.svs is a virtual slide that I just finished converting from
SVS to GoogleMaps format):
$ vipsheader UVa_LY03.svs
UVa_LY03.svs: 17296x26210 uchar, 3 bands, srgb, tiffload

Until now, I hadn't invoked *openslideload()* on the SIUE server, but I use
it routinely on a Windows computer to extract the slide labels from SVS
files. To explicitly look for *openslideload,* I tried to extract the slide
label from UVa_LY03.svs on the SIUE server, and the server confirmed that
it couldn't find *openslideload*:
$ openslideload --associated label UVa_LY03.svs UVa_LY03lbl.jpg
-bash: openslideload: command not found

As you suggested, not having *openslideload()* "installed" seems to be the
problem.  Previously, I've only installed libVips on a Windows computer,
and installing/configuring libVips on the SIUE server is something that my
colleague at SIUE probably has to do for me since I don't have the
necessary permissions (I've CC'd him on this reply).  Can the libVips that
already is loaded onto the SIUE server be reconfigured to invoke
openslideload(), or will my colleague need to remove the current instance
of libVips and download/install a new copy?

Again, many thanks for your help on this!

       - - Jim

**Comment by JRhodes-microscope** [comment]
I just realized that the command to explicitly look for openslideload() was
incorrect.  Here is the corrected command & result:
$ vips openslideload --associated label UVa_LY03.svs UVa_LY03lbl.jpg
vips: unknown action "openslideload"

Thanks again!

       - - Jim

**Comment by jcupitt** [comment]
It depends how you installed libvips. 

- If you built it yourself, you need to install `libopenslide-devel`, reconfigure and rebuild. 
- If you installed an RPM from somewhere, you'll need to look into how that RPM was built and see if they have one that does include openslide support.
- You could also consider installing docker, running something like ubuntu or debian stable in that, and using their libvips, which does include openslide.

**Comment by JRhodes-microscope** [comment]
John:

Just a short note to close-the-circle and thank you again for your help.
Your suggestions regarding our problem with color misrepresentation when
dzsave was used to convert "YUV" SVS slides accurately diagnosed and solved
the problem.  Our conversion of this type of file is up-and-running.
Again, MANY THANKS!!!

      - - Jim
